### PR TITLE
Add Java Properties language for .properties extension

### PR DESCRIPTION
**Pull request by eager**
This pull adds a Java Properties language for `.properties`.

## Description

I noticed that our some of our project’s `.properties` files had incorrect syntax highlighting after escaped `#`:

Screen shot of incorrect syntax highlighting

Because INI and Java Properties are very similiar—Codemirror’s `properties` mode appears to be a super-set of both—the disambiguation heuristic is somewhat complex. It replicates an outer-`if` block (`[key]=[value]`) that will match `INI`, unless the file also matches Java Properties’ comment markers (`^[#!]`), similar to the following code:
```ruby
if /^[^#!;][^=]*=/.match(data)
  # Match all key=value
  if /^[;\[]/.match(data)
    # Match INI comment token or sections
    Language["INI"]
  elsif /^[#!]/.match(data)
    # Match Java comment tokens
    Language["Java Properties"]
  else
    # Assume key=value is INI, if no comment tokens
    Language["INI"]
  end
elsif /^[^#!][^:]*:/.match(data)
  # Match key:value
  Language["Java Properties"]
end
```

## Checklist:

- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      
      -  https://github.com/search?o=desc&q=extension%3Aproperties+size%3A%3E1000+and+NOT+jdk&s=indexed&type=Code
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [INI/ms.properties: fnp/pylucene](https://github.com/fnp/pylucene/blob/fb16ac375de5479dec3919a5559cda02c899e387/samples/LuceneInAction/data/technology/computers/programming/education/ms.properties)
      - [INI/defaults.properties: phingofficial/phing](https://github.com/phingofficial/phing/blob/b4532bb8146841e5f60c1378e7d572bf8692a194/classes/phing/tasks/defaults.properties)
      - [Java Properties/libraries.properties: dustin/java-memcached-client](https://github.com/dustin/java-memcached-client/blob/c232307ad8e0c7ccc926e495dd7d5aad2d713318/ivy/libraries.properties)
      - [Java Properties/sounds.properties: spullara/maptool](https://github.com/spullara/maptool/blob/43a8e57b3e98e8b48949e73d943980e1190b217d/src/net/rptools/maptool/client/sounds.properties)
    - Sample license(s):
      - [phingofficial/phing: LGPL-3.0](https://github.com/phingofficial/phing/blob/b4532bb8146841e5f60c1378e7d572bf8692a194/LICENSE)
      - [dustin/java-memcached-client: MIT](https://github.com/dustin/java-memcached-client/blob/c232307ad8e0c7ccc926e495dd7d5aad2d713318/LICENSE.txt)
      - [spullara/maptool: Apache 2.0](https://github.com/spullara/maptool/blob/43a8e57b3e98e8b48949e73d943980e1190b217d/src/net/rptools/maptool/client/sounds.properties)
      - [fnp/pylucene: Apache 2.0](https://github.com/fnp/pylucene/blob/fb16ac375de5479dec3919a5559cda02c899e387/LICENSE)
  - [x] I have included a syntax highlighting grammar.
  - [x] I have included a change to the heuristics to distinguish my language from others using the same extension.
- [ ] **I am associating a language with a new file extension.**
  - [ ] The new extension is used in hundreds of repositories on GitHub.com
    - Search results for each extension:
      
      - https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3AFOOBAR+KEYWORDS+NOT+nothack
  - [ ] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [ ] **I am fixing a misclassified language**
  - [ ] I have included a new sample for the misclassified language:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [ ] **I am changing the source of a syntax highlighting grammar**
  
  - Old: https://github-lightshow.herokuapp.com/
  - New: https://github-lightshow.herokuapp.com/

- [ ] **I am adding new or changing current functionality**
  
  - [ ] I have added or updated the tests for the new or changed functionality.

### CODE REVIEW

**Comment by pchaigno** [comment]
Using @Alhadis' [Harvester script](https://github.com/Alhadis/Harvester), I collected 2913 sample urls (from the 19M total) from the search [`extension:properties`](https://github.com/search?utf8=%E2%9C%93&q=extension%3Aproperties+NOT+djfhdfkdnfk&type=) and downloaded the 615 corresponding repositories. According to Linguist, out of those 615 repositories with at least one `.properties` file, 226 are 100% Java, 457 have some Java (including the 226), and 158 don't contain any Java.

@eager Based on this analysis, I think we'll need to keep the `.properties` extension as an INI extension too, find a sample for INI, and add a heuristic rule to disambiguate the two.

**Comment by eager** [comment]
@pchaigno thanks for the review and pointer to Harvester. I had only been doing spot-checking of about a dozen results, so I’ll grab some more samples from there for disambiguation.

A few questions:

1. Are there project guidelines for how to handle ambiguous languages that surface during a samples search, but that aren‘t in the original scope?

I noticed in my initial spot-checking that there was some usage of `.properties` that were neither INI nor Java Properties, such as [this](https://github.com/openastexviewer/openastexviewer/blob/5b054fa8a12a9bd5f9a6394302ad2add0dd80ba5/src/lipophilicity.properties). My best guess is they are for the [OpenAstexViewer Scripting Language](http://openastexviewer.net/web/scripting.html).

2. Any sense of which language we should default to when it’s impossible to disambiguate? I don’t think it will be possible to disambiguate the most basic sample:
```ini
key = value
```

3. Would it help to remove `group: Java` from the language definition, so that if we lean toward Java Properties, for its more robust syntax highlighting, it doesn’t add misleading repo statistics?

**Comment by lildude** [comment]
Nudge.

**Comment by pchaigno** [comment]
Sorry for the delay @eager.

Could you add a test for the new heuristic rules in [`test_heuristics.rb`](https://github.com/github/linguist/blob/eeb5dcfff675b500005ac9da4fe0fc608dc0ae52/test/test_heuristics.rb)? We should be good to go after that.

File: lib/linguist/heuristics.rb [hunk 1]
```diff
@@ -392,6 +392,26 @@ def call(data)
...
+          Language["INI"]
+        elsif /^[#!]/.match(data)
+          # Match Java comment tokens
+          Language["Java Properties"]
+        else
+          # Assume key=value is INI
+          Language["INI"]
+        end
+      end
+      if /^[^#!][^:]*:/.match(data)
```
Comment (pchaigno) [inline lib/linguist/heuristics.rb]: Should this be an `elsif`? It's either `key=value` or `key:value`, right? I think that will fix the test failure.

**Comment by pchaigno** [comment]
@eager Did you get a chance to work again on this lately?

**Comment by eager** [comment]
@pchaigno apologies for disappearing on this for a while. Updated, with the test failure fixed.

**Pull request merged**